perl/lib/Gherkin/MarkdownTokenMatcher.pm

```diff
@@ -0,0 +1,346 @@
+    if ( $token->line->_trimmed_line_text =~ qr/$regex/ ) {
+        my $indent  = $token->line->indent + ( length($1) || 0 );
+        my $keyword = $2;
+        my $text    = $3;
+        $text =~ s/\s+$//;
+        my $keyword_type;
+        if ( exists $self->_keyword_types->{$keyword} ) {
+            # only set the keyword type if this is a step keyword
+            $keyword_type =
+              ( scalar( @{ $self->_keyword_types->{$keyword} } ) > 1 )
+              ? Cucumber::Messages::Step::KEYWORDTYPE_UNKNOWN
+              : $self->_keyword_types->{$keyword}->[0];
+        }
+        $self->_set_token_matched( $token, $token_type,
+            { indent => $indent, keyword => $keyword, text => $text, keyword_type => $keyword_type } );
+    }
+}
+
+sub _set_token_matched {
+    my ( $self, $token, $matched_type, $options ) = @_;
+    $options->{'items'} ||= [];
+    $token->matched_type($matched_type);
+
+    if ( defined $options->{'text'} ) {
+        chomp( $options->{'text'} );
+        $token->matched_text( $options->{'text'} );
+    }
+
+    $token->matched_keyword( $options->{'keyword'} )
+      if defined $options->{'keyword'};
+    $token->matched_keyword_type( $options->{'keyword_type'} )
+      if defined $options->{'keyword_type'};
+
+    if ( defined $options->{'indent'} ) {
+        $token->matched_indent( $options->{'indent'} );
+    } else {
+        $token->matched_indent( $token->line ? $token->line->indent : 0 );
+    }
+
+    $token->matched_items( $options->{'items'} )
+      if defined $options->{'items'};
+
+    $token->location->{'column'} = $token->matched_indent + 1;
+    $token->matched_gherkin_dialect( $self->dialect_name );
+}
+
+sub match_EOF {
+    my ( $self, $token ) = @_;
+    return unless $token->is_eof;
+    $self->_set_token_matched( $token, 'EOF' );
+}
+
+sub match_Empty {
+    my ( $self, $token ) = @_;
+    if (
+        $token->line->is_empty
+        || (   !$self->match_TagLine($token)
+            && !$self->match_FeatureLine($token)
+            && !$self->match_ScenarioLine($token)
+            && !$self->match_BackgroundLine($token)
+            && !$self->match_ExamplesLine($token)
+            && !$self->match_RuleLine($token)
+            && !$self->match_TableRow($token)
+            && !$self->match_Comment($token)
+            && !$self->match_Language($token)
+            && !$self->match_DocStringSeparator($token)
+            && !$self->match_EOF($token)
+            && !$self->match_StepLine($token) )
+      )
+    {
+        $self->_set_token_matched( $token, Empty => { indent => 0 } );
+    }
+}
+
+sub match_Comment {
+    my ( $self, $token ) = @_;
+    if (   $token->line->startswith('|')
+        && $self->_is_gfm_table_separator( $token->line->table_cells ) )
+    {
+        $self->_set_token_matched( $token,
+            Comment => { text => $token->line->line_text, indent => 0 } );
```

Comment:
@fperrad @mpkorstanje From what I understand from the JavaScript implementation, this is where the difference is: the JavaScript implementation sets a `token.matchedType` on all `match_XXX` functions *except* for on `match_Comment`. However, the Perl implementation *does* set it in this line by passing the "Comment =>" above. Try setting the `token.matchedType` to `TokenType.Comment` in the JS implementation and see if they now both fail the tests.

I *think* failing to set the `token.matchedType` simply causes comments to be ignored, which would be roughly the same as just not calling `_set_token_matched()` abov.